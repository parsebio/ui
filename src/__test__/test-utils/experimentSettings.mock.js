import _ from 'lodash';

import initialState from 'redux/reducers/experimentSettings/initialState';
import { analysisTools } from 'utils/constants';

const sampleifiedConfig = (sampleIds, configToReplicate) => {
  const result = sampleIds.reduce(
    (acum, sampleId) => {
      // eslint-disable-next-line no-param-reassign
      acum[sampleId] = configToReplicate;
      return acum;
    },
    {},
  );

  return result;
};

const generateProcessingConfigMock = (sampleIds) => ({
  classifier: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      prefiltered: false,
      filterSettings: {
        FDR: 0.1,
      },
    }),
  },
  cellSizeDistribution: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        minCellSize: 10800,
        binStep: 200,
      },
    }),
  },
  mitochondrialContent: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        method: 'absoluteThreshold',
        methodSettings: {
          absoluteThreshold: {
            maxFraction: 0.1,
            binStep: 200,
          },
        },
      },
    }),
  },
  numGenesVsNumUmis: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        regressionType: 'gam',
        regressionTypeSettings: {
          gam: {
            'p.level': 0.00009,
          },
        },
      },
    }),
  },
  doubletScores: {
    ...sampleifiedConfig(sampleIds, {
      auto: true,
      enabled: true,
      filterSettings: {
        probabilityThreshold: 0.2,
        binStep: 0.05,
      },
    }),
  },
  dataIntegration: {
    enabled: true,
    analysisTool: analysisTools.SEURAT,
    dataIntegration: {
      method: 'harmony',
      methodSettings: {
        harmony: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        seuratv4: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        fastmnn: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
        unisample: {
          numGenes: 2000,
          normalisation: 'logNormalize',
        },
      },
    },
    dimensionalityReduction: {
      method: 'rpca',
      numPCs: 30,
      excludeGeneCategories: [],
    },
  },
  configureEmbedding: {
    enabled: true,
    embeddingSettings: {
      method: 'umap',
      useSaved: false,
      methodSettings: {
        umap: {
          minimumDistance: 0.1,
          distanceMetric: 'euclidean',
        },
        tsne: {
          perplexity: 30,
          learningRate: 200,
        },
      },
    },
    clusteringSettings: {
      method: 'louvain',
      methodSettings: {
        louvain: {
          resolution: 0.5,
        },
      },
    },
  },
});

const generateExperimentSettingsMock = (sampleIds) => {
  const mockedProcessingConfig = {
    ...initialState.processing,
    ...generateProcessingConfigMock(sampleIds),
  };

  return {
    ...initialState,
    processing: mockedProcessingConfig,
    originalProcessing: _.cloneDeep(mockedProcessingConfig),
    info: {
      ...initialState.info,
      pipelineVersion: 1,
    },
  };
};

export default generateExperimentSettingsMock;
