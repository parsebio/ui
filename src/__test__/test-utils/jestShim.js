/**
 * This file is required for jest to work with '@cfaester/enzyme-adapter-react-18'
 * module.
 */
import {
  TextDecoder, TextEncoder,
} from 'util';

global.TextEncoder = TextEncoder;
global.TextDecoder = TextDecoder;
