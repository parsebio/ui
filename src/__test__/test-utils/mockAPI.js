import _ from 'lodash';
import backendStatusData from '__test__/data/backend_status.json';
import processingConfigData from '__test__/data/processing_config.json';
import mockDemoExperiments from '__test__/test-utils/mockData/mockDemoExperiments.json';
import { mockSecondaryAnalyses } from '__test__/data/secondaryAnalyses/secondary_analyses';
import mockSecondaryAnalysisStatusDefault from '__test__/data/secondaryAnalyses/secondary_analysis_status_default.json';
import fake from '__test__/test-utils/constants';
import mockAnalysisFiles from '__test__/data/secondaryAnalyses/secondary_analysis_files';
import downloadFromS3 from 'utils/work/downloadFromS3';

import {
  responseData,
} from '__test__/test-utils/mockData';
import { analysisTools } from 'const';

const cellSetsData = require('__test__/data/cell_sets.json');

const setupDownloadCellSetsFromS3Mock = (customCellSets = cellSetsData) => {
  downloadFromS3.mockImplementation((resource, signedUrl) => {
    // eslint-disable-next-line global-require
    if (signedUrl === 'mock-cellsets-signed-url') {
      return Promise.resolve(customCellSets);
    }
    return Promise.reject(new Error('Invalid signed URL'));
  });
};

const promiseResponse = (response, options = {}) => (
  Promise.resolve(new Response(response, options)));

const statusResponse = (code, body) => (
  Promise.resolve({
    status: code,
    body: JSON.stringify(body),
  })
);

const delayedResponse = (response, delay = 10000, options = {}) => new Promise((resolve) => {
  setTimeout(() => {
    resolve(new Response(response, options));
  }, delay);
});

const workerDataResult = (data) => Promise.resolve(_.cloneDeep(data));

const fetchWorkMock = (mockedResults) => ((experimentId, body) => {
  if (typeof mockedResults[body.name] === 'function') {
    return mockedResults[body.name]();
  }
  return workerDataResult(mockedResults[experimentId]);
});

const generateDefaultMockAPIResponses = (projectId, analysisTool = analysisTools.SEURAT) => ({
  [`experiments/${projectId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === projectId)),
  ),
  [`/v2/experiments/${projectId}$`]: () => promiseResponse(
    JSON.stringify(responseData.experiments.find(({ id }) => id === projectId)),
  ),
  [`experiments/${projectId}/processingConfig$`]: () => {
    const processingConfigDataClone = _.cloneDeep(processingConfigData);
    processingConfigDataClone.processingConfig.dataIntegration.analysisTool = analysisTool;

    return promiseResponse(
      JSON.stringify(processingConfigDataClone),
    );
  },
  [`experiments/${projectId}/cellSets$`]: () => {
    setupDownloadCellSetsFromS3Mock();
    return promiseResponse(
      JSON.stringify('mock-cellsets-signed-url'),
    );
  },
  [`experiments/${projectId}/backendStatus$`]: () => promiseResponse(
    JSON.stringify(backendStatusData),
  ),
  experiments$: () => promiseResponse(
    JSON.stringify(responseData.experiments),
  ),
  [`experiments/${projectId}/samples$`]: () => promiseResponse(
    JSON.stringify(responseData.samples[0]),
  ),
  '/v2/experiments/examples$': () => promiseResponse(
    JSON.stringify(mockDemoExperiments),
  ),
  'experiments/clone$': () => promiseResponse(
    JSON.stringify(fake.CLONED_EXPERIMENT_ID),
  ),
  [`/v2/workRequest/${projectId}`]: () => statusResponse(200, 'OK'),
  '/v2/secondaryAnalysis$': () => {
    const mockedMainSecondary = _.cloneDeep(mockSecondaryAnalyses[1]);
    mockedMainSecondary.id = projectId;

    return promiseResponse(
      JSON.stringify([
        mockedMainSecondary,
        ...mockSecondaryAnalyses,
      ]),
    );
  },
  [`/v2/secondaryAnalysis/${projectId}/executionStatus`]: () => promiseResponse(
    JSON.stringify(mockSecondaryAnalysisStatusDefault),
  ),
  [`/v2/secondaryAnalysis/${projectId}/files`]: (req) => {
    if (req.method === 'GET') {
      return promiseResponse(
        JSON.stringify(mockAnalysisFiles),
      );
    }

    if (req.method === 'POST') {
      return promiseResponse(
        JSON.stringify({ fileId: 'mockFileId' }),
      );
    }
  },
  [`/v2/projects/${projectId}/upload/.*/part/\\d+/signedUrl$`]: (req) => {
    const partNumber = req.url.match(/part\/(\d+)\/signedUrl/)[1];
    return promiseResponse(JSON.stringify(`mockSignedUrl${partNumber}`));
  },
  '/v2/cliUpload/token$': () => promiseResponse(JSON.stringify('token')),
});

const mockAPI = (apiMapping) => (req) => {
  const path = req.url;

  const key = _.find(Object.keys(apiMapping), (matcherStr) => new RegExp(matcherStr).test(path));

  if (!key) {
    return statusResponse({
      status: 404,
      body: `Path ${path} is undefined`,
    });
  }

  return apiMapping[key](req);
};

export default mockAPI;
export {
  generateDefaultMockAPIResponses,
  promiseResponse,
  statusResponse,
  delayedResponse,
  workerDataResult,
  fetchWorkMock,
  setupDownloadCellSetsFromS3Mock,
};
