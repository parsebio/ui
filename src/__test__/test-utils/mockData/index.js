import _ from 'lodash';
import { analysisTools } from 'utils/constants';
import generateMockSamples from './generateMockSamples';
import generateMockExperiments from './generateMockExperiments';
import generateMockProcessingConfig from './generateMockProcessingConfig';

// This file is a work in progress to generate mock data for the endpoints of the test.
// The generated data should be usable as response data to be sent via the mocked API in mockAPI.js
// The module should also expose the generated data that can be imported and used easily in tests

const responseData = {};

responseData.experiments = generateMockExperiments(2);

responseData.samples = [generateMockSamples(
  responseData.experiments[0],
  3,
)];

responseData.processingConfigSeurat = generateMockProcessingConfig(analysisTools.SEURAT, 3);
responseData.processingConfigScanpy = generateMockProcessingConfig(analysisTools.SCANPY, 3);

// Add samples to first experiment
const samples = responseData.samples[0];
const sampleIds = _.map(samples, 'id');
responseData.experiments[0].samplesOrder = sampleIds;

const { experiments, processingConfigSeurat, processingConfigScanpy } = responseData;

export {
  experiments,
  samples,
  processingConfigSeurat,
  processingConfigScanpy,
  responseData,
};
