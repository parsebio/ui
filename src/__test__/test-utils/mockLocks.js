const setupNavigatorLocks = () => {
  const mockLockRequest = jest.fn((lock, func) => func());
  global.navigator.locks = {
    request: mockLockRequest,
  };

  return mockLockRequest;
};

const teardownNavigatorLocks = () => {
  delete global.navigator.locks;
};

export { setupNavigatorLocks, teardownNavigatorLocks };
