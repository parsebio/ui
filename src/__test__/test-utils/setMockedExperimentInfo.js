import { updateExperimentInfo } from 'redux/actions/experimentSettings';

const setMockedExperimentInfo = async (experimentId, store) => {
  await store.dispatch(
    updateExperimentInfo({
      experimentId,
      experimentName: 'mockedName',
      sampleIds: ['1', '2'],
      pipelineVersion: 'v2',
      accessRole: 'owner',
    }),
  );
};

export default setMockedExperimentInfo;
