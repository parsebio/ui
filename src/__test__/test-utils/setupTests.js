import '@testing-library/jest-dom/extend-expect';
import preloadAll from 'jest-next-dynamic';

import Enzyme from 'enzyme';
import Adapter from '@cfaester/enzyme-adapter-react-18';

Enzyme.configure({ adapter: new Adapter() });

jest.mock('localforage');
jest.mock('utils/pushNotificationMessage');

// This is needed, because the unit tests don't register the Vitessce imports
jest.mock('components/data-exploration/DynamicVitessceWrappers', () => ({
  Scatterplot: () => 'Mocked Scatterplot',
  Heatmap: () => 'Mocked Heatmap',
}));

jest.mock('utils/work/downloadFromS3');

// The updated versions of react - dnd seem to not have a compatible version
// of the testing libraries anymore, so trying anything with the old way of setting up tests
// (with wrapWithTestBackend from react-dnd-test-utils or similar options) breaks.

jest.mock('react-dnd', () => {
  const mockBackend = { /* add methods as needed for your tests */ };
  const mockManager = {
    getBackend: jest.fn(() => mockBackend),
    getMonitor: jest.fn(() => ({
      subscribeToStateChange: jest.fn(),
      subscribeToOffsetChange: jest.fn(),
      isDragging: jest.fn(() => false),
      getItemType: jest.fn(),
      getItem: jest.fn(),
      canDragSource: jest.fn(),
      canDropOnTarget: jest.fn(),
      getDropResult: jest.fn(),
      didDrop: jest.fn(),
      getInitialClientOffset: jest.fn(),
      getInitialSourceClientOffset: jest.fn(),
      getSourceClientOffset: jest.fn(),
      getClientOffset: jest.fn(),
      getDifferenceFromInitialOffset: jest.fn(),
      isOverTarget: jest.fn(() => false),
      getTargetIds: jest.fn(),
    })),
    getActions: jest.fn(() => ({
      beginDrag: jest.fn(),
      publishDragSource: jest.fn(),
      hover: jest.fn(),
      drop: jest.fn(),
      endDrag: jest.fn(),
    })),
    getRegistry: jest.fn(() => ({
      addSource: jest.fn(),
      addTarget: jest.fn(),
      removeSource: jest.fn(),
      removeTarget: jest.fn(),
    })),
  };
  return {
    useDrag: () => [{ isDragging: false, handlerId: 'mock', draggedItem: null }, jest.fn(), jest.fn()],
    useDrop: () => [{ isOver: false, canDrop: false, handlerId: 'mock' }, jest.fn()],
    useDragDropManager: () => mockManager,
    DndProvider: ({ children }) => children,
  };
});

jest.mock('react-dnd-multi-backend', () => ({
  TouchTransition: {},
  HTML5toTouch: {},
  DndProvider: ({ children }) => children,
  createTransition: jest.fn(),
  default: {},
}));

jest.mock('rdndmb-html5-to-touch', () => ({
  default: {},
  HTML5toTouch: {},
}));

beforeAll(async () => {
  // Add stuff that needs to run once, before all tests
  await preloadAll();
});

beforeEach(async () => {
  // Add stuff that needs to run before each test
});

// setImmediate is not available in JSDOM, only on node
// https://github.com/prisma/prisma/issues/8558#issuecomment-1006100001
global.setImmediate = global.setImmediate || ((fn, ...args) => global.setTimeout(fn, 0, ...args));
