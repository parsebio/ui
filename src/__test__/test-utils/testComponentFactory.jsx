/* eslint-disable react/jsx-props-no-spreading */
import React from 'react';
import _ from 'lodash';

const createTestComponentFactory = (Component, defaultProps = {}) => (customProps = {}) => {
  // Merge is given an empty object so that it always
  // return a new object without mutating defaultProps
  const props = _.merge(
    {},
    defaultProps,
    customProps,
  );

  return <Component {...props} />;
};

export default createTestComponentFactory;
