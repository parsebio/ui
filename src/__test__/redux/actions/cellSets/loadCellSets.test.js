import _ from 'lodash';
import configureStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import fetchMock, { enableFetchMocks } from 'jest-fetch-mock';

import loadCellSets from 'redux/actions/cellSets/loadCellSets';
import initialState from 'redux/reducers/cellSets/initialState';

import '__test__/test-utils/setupTests';
import { CELL_SETS_ERROR, CELL_SETS_LOADED, CELL_SETS_LOADING } from 'redux/actionTypes/cellSets';
import mockAPI, { generateDefaultMockAPIResponses } from '__test__/test-utils/mockAPI';

enableFetchMocks();
const mockStore = configureStore([thunk]);
const experimentId = '1234';

const defaultResponses = generateDefaultMockAPIResponses(experimentId);

describe('loadCellSets action', () => {
  const experimentSettings = {
    info: {
      sampleIds: [],
    },
  };

  beforeEach(() => {
    fetchMock.resetMocks();
    fetchMock.mockIf(/.*/, mockAPI(defaultResponses));
  });

  it('Does not dispatch on normal operation', async () => {
    const store = mockStore({
      cellSets: {
        initialLoadPending: false, loading: false, error: false, updatingClustering: false,
      },
      experimentSettings,
    });

    await store.dispatch(loadCellSets(experimentId));
    expect(store.getActions().length).toEqual(0);
  });

  it('Dispatches on force reload', async () => {
    const store = mockStore({ cellSets: { loading: false, error: false }, experimentSettings });
    await store.dispatch(loadCellSets(experimentId, true));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_LOADED]);
  });

  it('Dispatches on initial load pending false and error state', async () => {
    const store = mockStore({
      cellSets: { initialLoadPending: false, loading: false, error: true },
      experimentSettings,
    });

    await store.dispatch(loadCellSets(experimentId));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_LOADED]);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches a loaded action when run with the initial state.', async () => {
    const store = mockStore({ cellSets: initialState, experimentSettings });
    await store.dispatch(loadCellSets(experimentId));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_LOADED]);

    expect(store.getActions()).toMatchSnapshot();
  });

  it('Dispatches an error condition if fetch fails', async () => {
    const store = mockStore({ cellSets: initialState, experimentSettings });

    fetchMock.resetMocks();
    fetchMock.mockReject(new Error('some weird error that happened'));

    await store.dispatch(loadCellSets(experimentId));

    expect(_.map(store.getActions(), 'type')).toEqual([CELL_SETS_LOADING, CELL_SETS_ERROR]);
    expect(store.getActions()).toMatchSnapshot();
  });

  it('Uses V2 URL when using API version V2', async () => {
    const store = mockStore({ cellSets: initialState, experimentSettings });
    await store.dispatch(loadCellSets(experimentId));

    const fetchUrl = fetchMock.mock.calls[0][0];

    expect(fetchUrl).toEqual('http://localhost:3000/v2/experiments/1234/cellSets');
  });
});
