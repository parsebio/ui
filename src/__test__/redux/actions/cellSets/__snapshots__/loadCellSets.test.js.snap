// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`loadCellSets action Dispatches a loaded action when run with the initial state. 1`] = `
[
  {
    "type": "cellSets/loading",
  },
  {
    "payload": {
      "data": [
        {
          "children": [
            {
              "cellIds": [
                1,
                2,
                3,
                4,
                5,
                6,
                7,
                8,
                9,
                10,
                11,
                12,
                16,
                17,
                18,
                19,
                20,
                21,
                22,
                23,
                24,
                25,
                26,
                27,
                28,
                29,
                30,
              ],
              "color": "#e377c2",
              "key": "louvain-0",
              "name": "Cluster 0",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                31,
                32,
                33,
                34,
                35,
                36,
                37,
                38,
                39,
                40,
                41,
                42,
                43,
                47,
                48,
                49,
                50,
                51,
                52,
                53,
                54,
                55,
              ],
              "color": "#8c564b",
              "key": "louvain-1",
              "name": "Cluster 1",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                57,
                58,
                59,
                60,
              ],
              "color": "#d62728",
              "key": "louvain-2",
              "name": "Cluster 2",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                61,
                62,
                64,
                65,
              ],
              "color": "#2ca02c",
              "key": "louvain-3",
              "name": "Cluster 3",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                66,
                68,
                69,
                70,
              ],
              "color": "#ff7f0e",
              "key": "louvain-4",
              "name": "Cluster 4",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                71,
                72,
                73,
                75,
              ],
              "color": "#1f77b4",
              "key": "louvain-5",
              "name": "Cluster 5",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                76,
                77,
                79,
                80,
              ],
              "color": "#f8e71c",
              "key": "louvain-6",
              "name": "Cluster 6",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                81,
                82,
                84,
                85,
              ],
              "color": "#3957ff",
              "key": "louvain-7",
              "name": "Cluster 7",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                88,
                89,
                90,
              ],
              "color": "#d3fe14",
              "key": "louvain-8",
              "name": "Cluster 8",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                91,
                92,
                93,
              ],
              "color": "#c9080a",
              "key": "louvain-9",
              "name": "Cluster 9",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                94,
                95,
              ],
              "color": "#fec7f8",
              "key": "louvain-10",
              "name": "Cluster 10",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                96,
                97,
              ],
              "color": "#0b7b3e",
              "key": "louvain-11",
              "name": "Cluster 11",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                98,
                99,
              ],
              "color": "#0bf0e9",
              "key": "louvain-12",
              "name": "Cluster 12",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                100,
              ],
              "color": "#c203c8",
              "key": "louvain-13",
              "name": "Cluster 13",
              "rootNode": false,
              "type": "cellSets",
            },
          ],
          "key": "louvain",
          "name": "fake louvain clusters",
          "rootNode": true,
          "type": "cellSets",
        },
        {
          "children": [],
          "key": "scratchpad",
          "name": "Custom cell sets",
          "rootNode": true,
          "type": "cellSets",
        },
        {
          "children": [
            {
              "cellIds": [
                1,
                2,
                3,
                4,
                5,
                6,
                7,
                8,
                9,
                10,
                11,
                12,
                13,
                14,
                15,
                16,
                17,
                18,
                19,
                20,
                21,
                22,
                23,
                24,
                25,
                26,
                27,
                28,
                29,
                30,
                31,
                32,
                33,
                34,
                35,
                36,
                37,
                38,
                39,
                40,
              ],
              "color": "#8c564b",
              "key": "b62028a1-ffa0-4f10-823d-93c9ddb88898",
              "name": "KO",
            },
            {
              "cellIds": [
                41,
                42,
                43,
                44,
                45,
                46,
                47,
                48,
                49,
                50,
                51,
                52,
                53,
                54,
                55,
                56,
                57,
                58,
                59,
                60,
                61,
                62,
                63,
                64,
                65,
                66,
                67,
                68,
                69,
                70,
                71,
                72,
                73,
                74,
                75,
              ],
              "color": "#d62728",
              "key": "ab568662-27fa-462c-9435-625594341314",
              "name": "WT1",
            },
            {
              "cellIds": [
                76,
                77,
                78,
                79,
                80,
                81,
                82,
                83,
                84,
                85,
                86,
                87,
                88,
                89,
                90,
                91,
                92,
                93,
                94,
                95,
                96,
                97,
                98,
                99,
                100,
              ],
              "color": "#e377c2",
              "key": "98c7a0ed-d086-4df8-bf94-a9ee6edb793f",
              "name": "WT2",
            },
          ],
          "key": "sample",
          "name": "Samples",
          "rootNode": true,
          "type": "metadataCategorical",
        },
        {
          "children": [
            {
              "cellSetKeys": [
                "98c7a0ed-d086-4df8-bf94-a9ee6edb793f",
              ],
              "color": "#8c564b",
              "key": "Track_1-KMeta",
              "name": "KMeta",
            },
            {
              "cellSetKeys": [
                "b62028a1-ffa0-4f10-823d-93c9ddb88898",
                "ab568662-27fa-462c-9435-625594341314",
              ],
              "color": "#d62728",
              "key": "Track_1-WMetaT",
              "name": "WMetaT",
            },
          ],
          "key": "Track_1",
          "name": "Track_1",
          "rootNode": true,
          "type": "metadataCategorical",
        },
      ],
      "experimentId": "1234",
    },
    "type": "cellSets/loaded",
  },
]
`;

exports[`loadCellSets action Dispatches an error condition if fetch fails 1`] = `
[
  {
    "type": "cellSets/loading",
  },
  {
    "payload": {
      "error": "We couldn't get the list of cell sets.",
    },
    "type": "cellSets/error",
  },
]
`;

exports[`loadCellSets action Dispatches on initial load pending false and error state 1`] = `
[
  {
    "type": "cellSets/loading",
  },
  {
    "payload": {
      "data": [
        {
          "children": [
            {
              "cellIds": [
                1,
                2,
                3,
                4,
                5,
                6,
                7,
                8,
                9,
                10,
                11,
                12,
                16,
                17,
                18,
                19,
                20,
                21,
                22,
                23,
                24,
                25,
                26,
                27,
                28,
                29,
                30,
              ],
              "color": "#e377c2",
              "key": "louvain-0",
              "name": "Cluster 0",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                31,
                32,
                33,
                34,
                35,
                36,
                37,
                38,
                39,
                40,
                41,
                42,
                43,
                47,
                48,
                49,
                50,
                51,
                52,
                53,
                54,
                55,
              ],
              "color": "#8c564b",
              "key": "louvain-1",
              "name": "Cluster 1",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                57,
                58,
                59,
                60,
              ],
              "color": "#d62728",
              "key": "louvain-2",
              "name": "Cluster 2",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                61,
                62,
                64,
                65,
              ],
              "color": "#2ca02c",
              "key": "louvain-3",
              "name": "Cluster 3",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                66,
                68,
                69,
                70,
              ],
              "color": "#ff7f0e",
              "key": "louvain-4",
              "name": "Cluster 4",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                71,
                72,
                73,
                75,
              ],
              "color": "#1f77b4",
              "key": "louvain-5",
              "name": "Cluster 5",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                76,
                77,
                79,
                80,
              ],
              "color": "#f8e71c",
              "key": "louvain-6",
              "name": "Cluster 6",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                81,
                82,
                84,
                85,
              ],
              "color": "#3957ff",
              "key": "louvain-7",
              "name": "Cluster 7",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                88,
                89,
                90,
              ],
              "color": "#d3fe14",
              "key": "louvain-8",
              "name": "Cluster 8",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                91,
                92,
                93,
              ],
              "color": "#c9080a",
              "key": "louvain-9",
              "name": "Cluster 9",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                94,
                95,
              ],
              "color": "#fec7f8",
              "key": "louvain-10",
              "name": "Cluster 10",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                96,
                97,
              ],
              "color": "#0b7b3e",
              "key": "louvain-11",
              "name": "Cluster 11",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                98,
                99,
              ],
              "color": "#0bf0e9",
              "key": "louvain-12",
              "name": "Cluster 12",
              "rootNode": false,
              "type": "cellSets",
            },
            {
              "cellIds": [
                100,
              ],
              "color": "#c203c8",
              "key": "louvain-13",
              "name": "Cluster 13",
              "rootNode": false,
              "type": "cellSets",
            },
          ],
          "key": "louvain",
          "name": "fake louvain clusters",
          "rootNode": true,
          "type": "cellSets",
        },
        {
          "children": [],
          "key": "scratchpad",
          "name": "Custom cell sets",
          "rootNode": true,
          "type": "cellSets",
        },
        {
          "children": [
            {
              "cellIds": [
                1,
                2,
                3,
                4,
                5,
                6,
                7,
                8,
                9,
                10,
                11,
                12,
                13,
                14,
                15,
                16,
                17,
                18,
                19,
                20,
                21,
                22,
                23,
                24,
                25,
                26,
                27,
                28,
                29,
                30,
                31,
                32,
                33,
                34,
                35,
                36,
                37,
                38,
                39,
                40,
              ],
              "color": "#8c564b",
              "key": "b62028a1-ffa0-4f10-823d-93c9ddb88898",
              "name": "KO",
            },
            {
              "cellIds": [
                41,
                42,
                43,
                44,
                45,
                46,
                47,
                48,
                49,
                50,
                51,
                52,
                53,
                54,
                55,
                56,
                57,
                58,
                59,
                60,
                61,
                62,
                63,
                64,
                65,
                66,
                67,
                68,
                69,
                70,
                71,
                72,
                73,
                74,
                75,
              ],
              "color": "#d62728",
              "key": "ab568662-27fa-462c-9435-625594341314",
              "name": "WT1",
            },
            {
              "cellIds": [
                76,
                77,
                78,
                79,
                80,
                81,
                82,
                83,
                84,
                85,
                86,
                87,
                88,
                89,
                90,
                91,
                92,
                93,
                94,
                95,
                96,
                97,
                98,
                99,
                100,
              ],
              "color": "#e377c2",
              "key": "98c7a0ed-d086-4df8-bf94-a9ee6edb793f",
              "name": "WT2",
            },
          ],
          "key": "sample",
          "name": "Samples",
          "rootNode": true,
          "type": "metadataCategorical",
        },
        {
          "children": [
            {
              "cellSetKeys": [
                "98c7a0ed-d086-4df8-bf94-a9ee6edb793f",
              ],
              "color": "#8c564b",
              "key": "Track_1-KMeta",
              "name": "KMeta",
            },
            {
              "cellSetKeys": [
                "b62028a1-ffa0-4f10-823d-93c9ddb88898",
                "ab568662-27fa-462c-9435-625594341314",
              ],
              "color": "#d62728",
              "key": "Track_1-WMetaT",
              "name": "WMetaT",
            },
          ],
          "key": "Track_1",
          "name": "Track_1",
          "rootNode": true,
          "type": "metadataCategorical",
        },
      ],
      "experimentId": "1234",
    },
    "type": "cellSets/loaded",
  },
]
`;
